Update content to use OSCAL v1.1.1 models

aj-stein-nist opened the pull request:
# Committer Notes

This updates oscal-content `make`-based orchestration and touches up catalogs for #116. Guidance on these changes, future development, and future governance are further detailed in usnistgov/OSCAL#1947.

Example content with the updated tooling work and simulating this branch being merged can be found on my work (the last commit changing the `HOME_REPO` variable to work on this branch has been dropped, but was used to complete this final testing).

https://github.com/aj-stein-nist/oscal-content/commit/c8a5374814e1917c527f567568cbbf200740b344#diff-ecf37eddbbe732cfc8cb391c407ae4ad8f00500d1590cc65bb08569c80697ff2

### All Submissions:

- [x] Have you followed the guidelines in our [Contributing](https://github.com/usnistgov/oscal-content/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/usnistgov/oscal-content/pulls) for the same update/change?
- ~Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\] (NOTE: Will do after review to aid review process and properly document many changes, big and small, potentially will not squash.)~ In this case, I believe it to be important that we keep all the commits for team understanding of significant changes in the repo. I will not squash these.
- [x] Do all automated CI/CD checks pass?

### Changes to Core Features:

- ~Have you added an explanation of what your changes do and why you'd like us to include them?~
- ~Have you written new tests for your core changes, as applicable?~
- ~Have you included examples of how to use your new feature(s)?~

iMichaela commented:
@aj-stein-nist Is the title of the PR wrong? It lists OSCAL 1.0.0

aj-stein-nist commented:
> @aj-stein-nist Is the title of the PR wrong? It lists OSCAL 1.0.0

Woops, fixed. :-)

aj-stein-nist commented:
This is important and still needs to be finished. I will continue work on this in the first half of Sprint 74 and intend to finish it as it is an important prerequisite for many upcoming changes given the tentative roadmap in usnistgov/OSCAL#1867.

aj-stein-nist commented:
OK testing progress is being made, we are close to the finish line on this one.

https://github.com/aj-stein-nist/oscal-content/actions/runs/6436935601/job/17481254664

iMichaela commented on .github/workflows/content-artifacts.yml:
```diff
@@ -9,17 +9,14 @@ on:
+  HOME_REPO: usnistgov/OSCAL
+  # With the default GEN_CONTENT_DIR, the resulting catalogs, profiles, and
+  # other example content will end up in the build/generated directory. For
+  # checking in finalized content, it go into build/.., the top-level directory
```
finalize content, it **goes** into build/

aj-stein-nist replied:
I think we will need to meet as a team to understand this, it does by default go into `build/generated` in local development only by default, but to avoid significant churn and confusion during final deploy and to match structure of the existing repo (and not cause a really frightful commit where the whole repo gets reorganized at the same time), it goes to `build/..`, this is short-hand for "the directory above, build, AKA top-level directory of the repo." If we want further discussion, we can set this up during or after review time.

iMichaela replied:
My comment was pointing to a grammar error, not to any logistics. The grammar error is still there :
 "# checking in finalized content, **_it go into_** build/.., the top-level directory"
should be:
# checking in finalized content, **_it goes into_** build/.., the top-level directory

aj-stein-nist replied:
Ohhh, ok. Sometimes the missing context and just the typo fix with bold alludes, especially with multiples inlcuded. My bad. That can definitely be fixed.

aj-stein-nist replied:
Corrected in https://github.com/usnistgov/oscal-content/commit/87becb621d1be14d139b8ce41497aed11dac0164.

iMichaela commented:
@aj-stein-nist - the Title indicates content update to OSCAL v 1.1.1 but the content has been updated to OSCAL 1.1.0

iMichaela commented:
@aj-stein-nist - can you add the changes of the PR #82 to this one, or do you prefer that I will rebase PR #82 after you are done with this one and merge it separately?

aj-stein-nist commented:
> @aj-stein-nist - the Title indicates content update to OSCAL v 1.1.1 but the content has been updated to OSCAL 1.1.0

The submodule was relocated and updated to [`f24dd56`](https://github.com/usnistgov/OSCAL/tree/f24dd56d5569ade8489924cf6fc2640dc297bfbe), which is the current v1.1.1 tag. I sync my test on fork back to this branch Friday or earlier today.

aj-stein-nist commented:
> @aj-stein-nist - can you add the changes of the PR #82 to this one, or do you prefer that I will rebase PR #82 after you are done with this one and merge it separately?

I would prefer to handle this after I am done and rebase 82 following the completion of this work.

aj-stein-nist commented:
Woops comment clean up led to me munging a GHA actions version by commit reference. Re-running.

aj-stein-nist commented:
Current, the last commit pointing to my "HOME_REPO" is to properly simulate, if merged, how this updated `make` and GitHub Actions automation would work. Once approved the last commit on the stack will be removed. Per standup discussion today, there are a couple of open questions we ought to consider.

1. For now (I will humor in the future a different approach), we are automatically committing source artifacts in `src` "up the tree" much like we had done with the usnistgov/OSCAL repo, causing frequent conflicts and churn for staff developers. Do we want to keep the auto-committing at all, and instead have developers control release prep and merge?
1. I may appear to be having issues with `hashFiles()` function working correctly to key the OSCAL build dependencies (so `build/oscal/build`, not the build directory itself. I may be misreading the logging for cache misses though, more to follow.
1. I am currently troubleshooting with team if we can increase concurrency with (`-jN` where is the number of `N` concurrent jobs, like `-j2` seems to be the max for GHA). I am working with @nikitawootten-nist and others familiar with Makefiles to sort that out.

More to follow.

wendellpiez reviewed (changes requested):
Why are we changing `/metadata/oscal-version`? I do not see anything that would invalidate the instance against the 1.0.0 model - the instances remain valid to any 1.x schema until they use a feature not available in 1.0.0.

(Backward compatibility is *extremely* expensive especially when measured in opportunity costs. Declaring what is the most restrictive model to which you conform is good manners. The backward-compatibility contract ensures that is generally the oldest.)

FWIW I am not saying the declared `oscal-version` should not be changed, just that no rationale has been offered and it is not a no-brainer.

Also it appears that `/*/xsl:schemaLocation` has not been updated in instances - shouldn't that be removed entirely?

Happy to discuss these issues to dispatch quickly.

wendellpiez commented:
Lemme add ... I *do* think that the instances should be validated to the appropriate latest-greatest schemas and indeed this might be noted somewhere....

aj-stein-nist commented:
> Lemme add ... I _do_ think that the instances should be validated to the appropriate latest-greatest schemas and indeed this might be noted somewhere....

I was typing up my rationale, but this is the reason we ought to do it. I do not mind documenting it, but the only reason it has fallen behind was a significant amount of tech debt pulling up to the recent version. I was actually asking some of the above questions in https://github.com/usnistgov/oscal-content/pull/204#issuecomment-1755882436 to coincide with the publication of an ADR on this and other topics.

aj-stein-nist commented:
> (Backward compatibility is _extremely_ expensive especially when measured in opportunity costs. Declaring what is the most restrictive model to which you conform is good manners. The backward-compatibility contract ensures that is generally the oldest.)
>
> FWIW I am not saying the declared `oscal-version` should not be changed, just that no rationale has been offered and it is not a no-brainer.

If we believe what we say about our models and those guarantees and minor patch releases fix bugs and potentially add features that are compatible, is that not acceptable to also declare that with `oscal-version`?
 
> Also it appears that `/*/xsl:schemaLocation` has not been updated in instances - shouldn't that be removed entirely?

I would want to adjust that before this is finalized, barring other issues that need to be discussed before merge, and the import discussion of which version to use fro `xml-model` or another instruction, the location matters. I am sorry I did not include that in the list of pending concerns that were more tactical in nature.

But these are good questions, thanks for pointing them out.

aj-stein-nist commented:
> > @aj-stein-nist - the Title indicates content update to OSCAL v 1.1.1 but the content has been updated to OSCAL 1.1.0
> 
> The submodule was relocated and updated to [`f24dd56`](https://github.com/usnistgov/OSCAL/tree/f24dd56d5569ade8489924cf6fc2640dc297bfbe), which is the current v1.1.1 tag. I sync my test on fork back to this branch Friday or earlier today.

I now realize, @iMichaela you may have been referring to the `oscal-version` of the documents (which was `1.1.0` until a few minutes ago). That is correct, I noticed this during more reviews and touch-ups today after this message. Either way, The original commit in this pull request is now rebased to reflect the correct version, `1.1.1`, in https://github.com/usnistgov/oscal-content/pull/204/commits/fd7367d0f6e5b62e1bb100fe7236bd7dfa860ce6.

aj-stein-nist commented:
> Also it appears that `/*/xsl:schemaLocation` has not been updated in instances - shouldn't that be removed entirely?

While cleaning up this PR I decided now to take this sooner and for now pointed to `xml-model` for validation for the latest release in https://github.com/usnistgov/oscal-content/pull/204/commits/8793427d2657115a2b8bf4c0a6adb1a1c38425ac.

iMichaela commented:
> > > @aj-stein-nist - the Title indicates content update to OSCAL v 1.1.1 but the content has been updated to OSCAL 1.1.0
> > 
> > 
> > The submodule was relocated and updated to [`f24dd56`](https://github.com/usnistgov/OSCAL/tree/f24dd56d5569ade8489924cf6fc2640dc297bfbe), which is the current v1.1.1 tag. I sync my test on fork back to this branch Friday or earlier today.
> 
> I now realize, @iMichaela you may have been referring to the `oscal-version` of the documents (which was `1.1.0` until a few minutes ago). That is correct, I noticed this during more reviews and touch-ups today after this message. Either way, The original commit in this pull request is now rebased to reflect the correct version, `1.1.1`, in [fd7367d](https://github.com/usnistgov/oscal-content/commit/fd7367d0f6e5b62e1bb100fe7236bd7dfa860ce6).

Indeed.  Thank you.

nikitawootten-nist commented on build/Makefile:
```diff
@@ -0,0 +1,169 @@
+	{ printf "\033[32m%-30s\033[0m %s\n", $$1, $$2 }'
+
+.PHONY: all
+all: \
+	dependencies             \
+	build                    \
+	copy-readmes             \
+	copy-xml-content         \
+	resolve-xml-profiles     \
+	convert-min-json-content \
+	reformat-json-content    \
+	convert-yaml-content     \
+	validate-xml-content     \
+	validate-json-content    \
+	validate-yaml-content ## Run all steps for content preparation
```
NB: A limitation of the `help` target is that it assumes targets are all on one line. I like the readability afforded by your multiline target, however it will not be printed by `make help` in its current form :(. I leave this trade-off to your judgement.

aj-stein-nist replied:
Ohhh, I hadn't even noticed this. 😢 That can be adjusted, I just skimmed the sub-targets. I shake my fist at the `make`-makers.

nikitawootten-nist replied:
Unfortunately it is me you should be shaking your fist at me, the `help` target just does not expect multi-line target dependencies. It's possible you can adjust the `grep` and `awk` commands, but that sounds like it could be more trouble than it's worth.

aj-stein-nist replied:
I'd rather not get too deep into that, I can remove that change in upcoming commits and rebasing (I think this had a dedicated cleanup commit, I will update on the change).

aj-stein-nist replied:
Done in https://github.com/usnistgov/oscal-content/pull/204/commits/caec68c96437997e1253467a5d8b499feb557d07.

aj-stein-nist commented:
Ok, to give a status update after I had missed standup and mentioned in chat, one of the issues we have to identify as a "deal breaker" or not is how the older versions of the XSLT profile resolver stripped the full path of the source profile in the resolved catalog, where the newer versions have a parameter to remote the whole path, file name included, or leave it as-is. So now we get examples like this.

```
      <prop name="resolution-tool"
             value="OSCAL Profile Resolver XSLT Pipeline OPRXP"/>
      <link href="file:/home/runner/work/oscal-content/oscal-content/build/../src/nist.gov/SP800-53/rev5/xml/NIST_SP-800-53_rev5_LOW-baseline_profile.xml"
             rel="source-profile"/>
```

(NOTE: from the penultimate generation of content https://github.com/aj-stein-nist/oscal-content/blob/dc5500ed9371b485fc21cb095d2fb9db6e2a1fd3/nist.gov/SP800-53/rev5/xml/NIST_SP-800-53_rev5_LOW-baseline-resolved-profile_catalog.xml#L11)

The alternative as-is would be this with the parameter (but I have removed it from the PR with testing):

```
      <prop name="resolution-tool"
             value="OSCAL Profile Resolver XSLT Pipeline OPRXP"/>
      <link href="profile"
             rel="source-profile"/>
```

@wendellpiez, am I understanding `hide-source-profile-uri` parameter correctly? I believe this is working as intended with current implementation, I just want to double-check here.

I want some feedback from the team if we want or need to alter this, as that property is informational.

nikitawootten-nist commented on build/Makefile:
```diff
@@ -0,0 +1,169 @@
+GEN_JSON_FILES:=$(subst xml,json,$(GEN_XML_FILES))
+GEN_MIN_JSON_FILES:=$(subst .json,-min.json,$(subst xml,json,$(GEN_XML_FILES)))
+GEN_YAML_FILES:=$(subst xml,yaml,$(GEN_XML_FILES))
+CURL_INSTALL_OPTS:=--silent --location
+JQ_BIN:=jq-linux-amd64
+JQ_PATH:=$(shell pwd)/jq
+JQ_INSTALL_VERSION:=jq-1.7
+JQ_DOWNLOAD_URL:=https://github.com/jqlang/jq/releases/download/$(JQ_INSTALL_VERSION)/$(JQ_BIN)
+JQ_INSTALL_COMMAND=curl $(CURL_INSTALL_OPTS) -o $(JQ_PATH) $(JQ_DOWNLOAD_URL) && chmod +x $(JQ_PATH)
+XMLLINT_PATH=/usr/bin/xmllint
+XMLLINT_INSTALL_COMMAND=sudo rm -rf /var/lib/apt/lists/* && sudo apt-get update && sudo apt-get install libxml2-utils
+YQ_BIN:=yq_linux_amd64
+YQ_PATH:=$(shell pwd)/yq
+YQ_INSTALL_VERSION:=v4.35.2
+YQ_DOWNLOAD_URL:=https://github.com/mikefarah/yq/releases/download/$(YQ_INSTALL_VERSION)/$(YQ_BIN)
+YQ_INSTALL_COMMAND:=curl $(CURL_INSTALL_OPTS) -o $(YQ_PATH) $(YQ_DOWNLOAD_URL) && chmod +x $(YQ_PATH)
```
B: I like this approach, however it unfortunately breaks in non-linux environments. The XML lint install also breaks on non-debian/ubuntu environments. Maybe we should just document these requirements in the readme as was done in the core oscal repo?

aj-stein-nist replied:
I thought the way you fix that for other Linux/Unixy environments it override the command on the command line, I actually tested this and was surprised how effective it is, but this means less, how can I say this, "elaborate" arguments when invoking `make` in GitHub Actions if we need to express them explicitly there, so I though it wasn't worth the headache.

Conversely, if you explain with the `XMLLINT_PATH` absolute target to where that file is, you can override to wherever you want. The only place I cannot see this work is Windows, am I mistaken?

aj-stein-nist replied:
Done in https://github.com/usnistgov/oscal-content/pull/204/commits/49a77648618e827838540385085228f90297ba44.

nikitawootten-nist commented on build/README.md:
````diff
@@ -0,0 +1,66 @@
+make convert-yaml-content
+```
+
+### Checks
+
+You can use the `make all` to validate generated artifacts in addition to other steps, or you can use specific target(s) to generate a specific artifact.
+
+```sh
+make validate-xml-content
+make validate-json-content
+make validate-yaml-content
+```
````
NB: May I suggest a `checks` make target that runs all validation commands? Ditto for artifact generation.

aj-stein-nist replied:
Done and done, those can be done. 👍

aj-stein-nist replied:
Done in https://github.com/usnistgov/oscal-content/pull/204/commits/cd1f7d71d8c4fa9921ad201e4894b54b22cb989c.

nikitawootten-nist reviewed (changes requested):
Looks great with one caveat, see blocking feedback below.

aj-stein-nist commented on build/README.md:
```diff
@@ -0,0 +1,66 @@
+
+If you want to run `make` commands outside of this directory, for example from [the top-level directory of this repository](..) cloned onto a developer workstation, use the `-C` argument, such as `make -C ./build all`.
+
+By default, the destination for all generated content is in the [`generated`](./generated) sub-directory. To generate content "inline" like the GitHub Actions CI/CD automation, you want to change the `GEN_CONTENT_DIR` relative to location of the `Makefile`. Either executing `make -C all GEN_CONTENT_DIR=..` from the top-level directory or executing `make -C build GEN_CONTENT_DIR=..` from this dire ...
```
I am embarrassed I miss this in a rush, these edits can be changed in a few short minutes.

aj-stein-nist replied:
Done in https://github.com/usnistgov/oscal-content/pull/204/commits/cd1f7d71d8c4fa9921ad201e4894b54b22cb989c.

aj-stein-nist commented on build/Makefile:
```diff
@@ -0,0 +1,167 @@
+.PHONY: resolve-xml-profiles
+resolve-xml-profiles: $(GEN_XML_RESOLVED_CATALOGS) ## Resolve OSCAL XML profiles for custom catalogs
+
+$(GEN_CONTENT_DIR)/%-resolved-profile_catalog.xml: $(SRC_DIR)/%_profile.xml
+	mkdir -p $(@D)
+	$(PROFILE_RESOLVER_RUNNER) $(GEN_CONTENT_DIR)/$*_profile.xml $(GEN_CONTENT_DIR)/$*-resolved-profile_catalog.xml $(PROFILE_RESOLVER_ARGS)
+	@sed -i'' -e "s|file:$(shell realpath $(GEN_CONTENT_DIR)/$*_profile.xml)|$(shell basename $*_profile.xml)|g" $(GEN_CONTENT_DIR)/$*-resolved-profile_catalog.xml	
```
Potentially blocking: I keep getting mixed up on the target syntax, but I noticed that we are not using the copied profiles into the `GEN_CONTENT_DIR` instead of those in `SRC_DIR`. I can change this for the `$(PROFILE_RESOLVER_RUNNER)` command but not the targets above the `mkdir` command. Ensure with the team during review I am doing this correctly.

aj-stein-nist replied:
@nikitawootten-nist I saw your review and I making changes, but to be clear: this is not an issue?

wendellpiez commented:
Thanks for the patience to read and contemplate - no action item here.

FWIW, updating the `oscal-version` actually says nothing about backward-compatibility other than implicitly, indeed it is rather the reverse - the policy says something about the use of the value, not the other way around.

If we wish to assure backward-compatibility *and* make the most complete claim, we could (for example) stipulate that `oscal-version` show the oldest version to which the document is known to be valid, while `<?xml-model?>` shows the latest.

Alternatively the policy could be to maintain both pinned to the latest - as an implicit claim that validation has at least been performed against the latest (something we should be doing *in any case*). That loses information regarding the oldest possible schema that can be used (or: which is supposed to have been used at time of origination), but you could argue that information isn't valuable.

Happy to discuss these data governace questions further as IMV they do not receive enough attention - *au contraire* too often decisions are made without discussing points of view and tradeoffs, leaving them to come back to bite us.

aj-stein-nist commented:
> Thanks for the patience to read and contemplate - no action item here.

OK.

> FWIW, updating the `oscal-version` actually says nothing about backward-compatibility other than implicitly, indeed it is rather the reverse - the policy says something about the use of the value, not the other way around.

This is why I intend to draft an ADR PR to usnistgov/OSCAL to talk about this usage of `oscal-version` separate of that.
 
> If we wish to assure backward-compatibility _and_ make the most complete claim, we could (for example) stipulate that `oscal-version` show the oldest version to which the document is known to be valid, while `<?xml-model?>` shows the latest.

We could, but I want an assurance that the example content in this repo is using the latest valid schema to ensure they work as the last form of integration and/or regression testing. Not having this assurance led to this PR and several other attempts being massive and unearthing edge case bugs because this method was not employed, and if it had been frequently issues would have been caught sooner. That will come up in the ADR.

> Alternatively the policy could be to maintain both pinned to the latest - as an implicit claim that validation has at least been performed against the latest (something we should be doing _in any case_). That loses information regarding the oldest possible schema that can be used (or: which is supposed to have been used at time of origination), but you could argue that information isn't valuable.

We can pin to one version without adding more complexity in infrastructure/automation and further obscuring to us (staff and downstream community) what is being tested. The oldest schema can be used, and documented that it can be done, but in the ADR I will have to address that this means we are not demonstrating we have validated current examples against the most recent schema. Through frequent questions in GitHub, Gitter, and casual interactions, this is a common source of confusion as communicated through code, and there is technical risk cited in the above reply. I will attempt to address this in the ADR.

> Happy to discuss these data governace questions further as IMV they do not receive enough attention - _au contraire_ too often decisions are made without discussing points of view and tradeoffs, leaving them to come back to bite us.

We have been here and I assume we will continue to with the rest of the team in the follow on ADR. I will update this comment with a link to that pull request later today.

wendellpiez commented:
Thanks AJ. So far so good. The Profile Resolver feature enhancement of course was a direct reflection of a bug or Issue raised by a community member (Gary G comes to mind as a possible offender, or a FedRAMP team member) -- as you know we tried our best not to ignore those reports as they came in, and the changes are the result (we didn't do them just for fun).

aj-stein-nist commented:
> Ok, to give a status update after I had missed standup and mentioned in chat, one of the issues we have to identify as a "deal breaker" or not is how the older versions of the XSLT profile resolver stripped the full path of the source profile in the resolved catalog, where the newer versions have a parameter to remote the whole path, file name included, or leave it as-is. So now we get examples like this.

To move this along, I have added the workaround in commit https://github.com/usnistgov/oscal-content/pull/204/commits/2e0323f4416f557123b99a4910213a5483a37ae7. I will file an issue in the OSCAL repo to request a change in the profile resolver code.

wendellpiez commented:
@aj-stein-nist @nikitawootten-nist Any reason we are not supporting post-processes for *any* generated content we are loading?

This could be more than filename rewriting. Whitespace rectification is an obvious such application that comes to mind. Our improvements in CI/CD might make introducing such a filter step much easier than until recently.

These could be XSLT or Python or anything else and could be more or less tightly integrated e.g. with XProc, or just piped.

aj-stein-nist commented:
> @aj-stein-nist @nikitawootten-nist Any reason we are not supporting post-processes for _any_ generated content we are loading?

As the person who made it this way: it makes it reproducible and consistent in all different environments. If we want to add post-processing steps, we add them to the pipeline, and that's easy enough. That is what I did with the use of `sed` in the `Makefile`. Is there something we are missing with that approach we cannot add later. Whitespace normalization in subsequent issues and PRs is exactly the kind of thing I would envision as another step in the `Makefile` target if and when we need it.

nikitawootten-nist commented on build/Makefile:
```diff
@@ -0,0 +1,167 @@
+YQ_PATH:=$(shell which yq || echo $(YQ_INSTALL_PATH))
+YQ_INSTALL_COMMAND:=curl $(CURL_INSTALL_OPTS) -o $(YQ_INSTALL_PATH) $(YQ_INSTALL_URL) && chmod +x $(YQ_INSTALL_PATH)
+
+$(YQ_PATH):
```
B: Correct behavior to make manual invocation possible on machines with `yq` installed:
```suggestion
$(YQ_INSTALL_PATH):
```

aj-stein-nist replied:
Corrected in https://github.com/usnistgov/oscal-content/commit/99669b4bd2603f0eea88a5d8a2ca2df76921b5d6.

nikitawootten-nist commented on build/Makefile:
```diff
@@ -0,0 +1,167 @@
+	@cat $(GEN_CONTENT_DIR)/$(subst yaml,json,$*).json | $(YQ_PATH) e -P - > $(GEN_CONTENT_DIR)/$(subst json,yaml,$*).yaml
+
+.PHONY: validate-yaml-content
+validate-yaml-content: $(GEN_YAML_FILES) ## Validate YAML files
+	npx --prefix $(NPM_PREFIX_DIR) ajv validate -s $(OSCAL_COMPLETE_JSON_SCHEMA) -c ajv-formats $(foreach file,$(GEN_YAML_FILES),-d $(file))
```
B: Include NPM install as a dependency for ajv validation
```suggestion
validate-yaml-content: $(GEN_YAML_FILES) $(NPM_PREFIX_DIR)/$(NPM_PKGS_DIR) ## Validate YAML files
	npx --prefix $(NPM_PREFIX_DIR) ajv validate -s $(OSCAL_COMPLETE_JSON_SCHEMA) -c ajv-formats $(foreach file,$(GEN_YAML_FILES),-d $(file))
```

aj-stein-nist replied:
Woops, good catch!

aj-stein-nist replied:
Corrected in https://github.com/usnistgov/oscal-content/commit/99669b4bd2603f0eea88a5d8a2ca2df76921b5d6.

nikitawootten-nist commented on build/Makefile:
```diff
@@ -0,0 +1,167 @@
+checks: validate-xml-content validate-json-content validate-yaml-content ## Check all content with schema and other validation methods
+
+.PHONY: validate-json-content
+validate-json-content: $(GEN_JSON_FILES) ## Validate JSON files
+	npx --prefix $(NPM_PREFIX_DIR) ajv validate -s $(OSCAL_COMPLETE_JSON_SCHEMA) -c ajv-formats $(foreach file,$(GEN_JSON_FILES),-d $(file))
```
B: NPM install dependancy
```suggestion
validate-json-content: $(GEN_JSON_FILES) $(NPM_PREFIX_DIR)/$(NPM_PKGS_DIR) ## Validate JSON files
	npx --prefix $(NPM_PREFIX_DIR) ajv validate -s $(OSCAL_COMPLETE_JSON_SCHEMA) -c ajv-formats $(foreach file,$(GEN_JSON_FILES),-d $(file))
```

aj-stein-nist replied:
Corrected in https://github.com/usnistgov/oscal-content/commit/99669b4bd2603f0eea88a5d8a2ca2df76921b5d6.

nikitawootten-nist commented on build/Makefile:
```diff
@@ -0,0 +1,167 @@
+.PHONY: convert-min-json-content
+convert-min-json-content: $(GEN_MIN_JSON_FILES) ## Convert examples from OSCAL XML to JSON
+
+.SECONDEXPANSION:
+$(GEN_CONTENT_DIR)/%-min.json: $(GEN_CONTENT_DIR)/$$(subst json,xml,%).xml
+	@mkdir -p $(@D)
+	$(XSLT_RUNNER) $(XML_JSON_CONVERTER_XSLT) $(GEN_CONTENT_DIR)/$(subst json,xml,$*).xml $(GEN_CONTENT_DIR)/$*-min.json
```
B: Getting errors here when running a parallel build. This target should depend on the converters explicitly

nikitawootten-nist commented on build/Makefile:
```diff
@@ -0,0 +1,167 @@
+	$(PROFILE_RESOLVER_RUNNER) $(GEN_CONTENT_DIR)/$*_profile.xml $(GEN_CONTENT_DIR)/$*-resolved-profile_catalog.xml $(PROFILE_RESOLVER_ARGS)
+	@sed -i'' -e "s|file:$(shell realpath $(GEN_CONTENT_DIR)/$*_profile.xml)|$(shell basename $*_profile.xml)|g" $(GEN_CONTENT_DIR)/$*-resolved-profile_catalog.xml	
+
+.PHONY: validate-xml-content
+validate-xml-content: $(GEN_XML_FILES) ## Validate XML files
+	@xmllint --schema $(OSCAL_COMPLETE_XML_SCHEMA) --noout $(GEN_XML_FILES)
```
B: This target should depend on the downstream oscal schema generation target

nikitawootten-nist reviewed (changes requested):
Looking much better, found a few problems which only show up on partial clean operations or parallel builds. See comments below.

aj-stein-nist commented:
Doing some more testing but the concurrency fix as I committed is still not working perfectly.

https://github.com/aj-stein-nist/oscal-content/actions/runs/6592140253/job/17912204423

Catching up on Makefile docs and tutorials while debugging locally and GHA to figure out if I did it in the wrong target location or this is not sufficient as discussed during pairing. 😭

aj-stein-nist commented:
It seems with https://github.com/usnistgov/oscal-content/pull/204/commits/c881404d821e758233f9b50ac80d53fa6abac5ff we are very close (thanks to @nikitawootten-nist for #219) but I still see a regression in the `sed` path transformation that is not being handled in the latest testing on my fork as of https://github.com/aj-stein-nist/oscal-content/commit/7392f08f3c2d11a36a293d59c3f6b5d83d17e3fc. Once that is wrapped up, this is ready.

aj-stein-nist commented:
I removed the testing commit to change the HOME_REPO, but this is ready for final review.

/cc @usnistgov/itl-oscal

nikitawootten-nist commented on build/Makefile:
```diff
@@ -0,0 +1,223 @@
+	SED_FLAGS:=-i ''
+endif
+
+$(GEN_CONTENT_DIR)/%-resolved-profile_catalog.xml: $(SRC_DIR)/%_profile.xml $(GEN_XML_COPIED)
```
NB: Small self-nitpick, I'd remove this extra target here. I tested and it all works without it.
```suggestion
$(GEN_CONTENT_DIR)/%-resolved-profile_catalog.xml: $(SRC_DIR)/%_profile.xml
```

nikitawootten-nist reviewed (approved):
All works, all approved. Small nitpick below but it can be safely ignored
